core/templates/components/graph-services/graph-layout.service.spec.ts

```diff
@@ -0,0 +1,1251 @@
+      height: 0.08,
+      width: 0.18000000000000002,
+      reachableFromEnd: true
+    },
+    Orphaned: {
+      depth: 5,
+      offset: 0,
+      reachable: false,
+      x0: 0.07250000000000001,
+      y0: 0.7933333333333333,
+      xLabel: 0.1625,
+      yLabel: 0.8333333333333333,
+      id: 'Orphaned',
+      label: 'Orphaned',
+      height: 0.08,
+      width: 0.18000000000000002,
+      reachableFromEnd: false
+    }
+  };
+  let links1: GraphLink[] = [
+    {
+      source: 'State1',
+      target: 'State1'
+    },
+    {
+      source: 'State1',
+      target: 'State2'
+    },
+    {
+      source: 'State1',
+      target: 'State3'
+    },
+    {
+      source: 'State1',
+      target: 'State5'
+    },
+    {
+      source: 'State1',
+      target: 'State6'
+    },
+    {
+      source: 'State1',
+      target: 'State7'
+    },
+    {
+      source: 'State1',
+      target: 'State8'
+    },
+    {
+      source: 'State1',
+      target: 'State9'
+    },
+    {
+      source: 'State2',
+      target: 'State4'
+    },
+    {
+      source: 'State3',
+      target: 'State4'
+    },
+    {
+      source: 'State9',
+      target: 'State8'
+    },
+    {
+      source: 'State8',
+      target: 'State4'
+    },
+  ];
+
+  let links2: GraphLink[] = [
+    {
+      source: 'State1',
+      target: 'State1'
+    },
+    {
+      source: 'State1',
+      target: 'State2'
+    },
+    {
+      source: 'State1',
+      target: 'State3'
+    },
+    {
+      source: 'State2',
+      target: 'State4'
+    },
+    {
+      source: 'State3',
+      target: 'State4'
+    },
+  ];
+
+  beforeEach(() => {
+    sgls = TestBed.inject(StateGraphLayoutService);
+  });
+
+  it('should create adjacency lists', () => {
+    let nodes: GraphNodes = {
+      State1: 'State1',
+      State2: 'State2',
+      State3: 'State3',
+      State4: 'State4',
+    };
+    let expectedAdjacencyLists = {
+      State1: ['State2', 'State3'],
+      State2: ['State4'],
+      State3: ['State4'],
+      State4: []
+    };
+
+    expect(sgls.getGraphAsAdjacencyLists(nodes, links2))
+      .toEqual(expectedAdjacencyLists);
+  });
+
+  it('should return indentation levels for a segment of nodes', () => {
+    let adjacencyLists = {
+      State1: ['State2', 'State3', 'State4'],
+      State2: ['State3', 'State4'],
+      State3: ['State4', 'State5'],
+      State4: ['State5'],
+      State5: []
+    };
+
+    let longestPathIds: string[] = [
+      'State1',
+      'State2',
+      'State3',
+      'State4',
+      'State5'
+    ];
+
+    expect(sgls.getIndentationLevels(adjacencyLists, longestPathIds)).toEqual(
+      [0, 0.5, 1, 0, 0]);
+
+    let shortestPathIds: string[] = [
+      'State1',
+      'State4',
+      'State5'
+    ];
+    expect(sgls.getIndentationLevels(adjacencyLists, shortestPathIds)).toEqual(
+      [0, 0, 0]);
+  });
+
+  it('should not return indentation level greater' +
+    ' than MAX_INDENTATION_LEVEL', () => {
+    let adjacencyLists = {
```

Comment:
Perhaps give a verbal or graphical explanation for why this graph would be expected to have high indentation for certain nodes.

Reply:
Done